core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java

```diff
@@ -202,4 +221,29 @@ public int hashCode() {
         result = 31 * result + (unit != null ? unit.hashCode() : 0);
         return result;
     }
+
+    @Override
+    public int compareTo(ByteSizeValue other) {
+        long unitValue = unit.toBytes(1);
+        long otherUnitValue = other.unit.toBytes(1);
+
+        ByteSizeUnit minUnit = unitValue < otherUnitValue ? unit : other.unit;
+        long thisValue = this.convert(minUnit);
+        long otherValue = other.convert(minUnit);
+
+        if (thisValue == otherValue) {
```

Comment:
> Is there a reason why we want this behavior? (I noticed the writeTo method also writes getBytes())

I'm not sure why we wanted it at some point, but we've gone away from that for `TimeValue` so I expect we should go away from it for `ByteSizeValue` eventually. I think maybe in another PR though rather than making this one bigger.

Yeah, the max bytes problem makes sense to me. I guess we should just make the equals and compareTo actually correct rather than have the rollover issue. I don't think we need equals or compareTo to be super efficient so I'd prefer if we just made the right even if it isn't fast. I suspect we've never hit this because `Long.MAX_VALUE` is so much bigger than anything anyone is particularly likely to provide as input. But, yeah, I'd love it if you made the comparator work properly for these values anyway.